src/System.IO.Pipelines.Networking.Tls/Internal/OpenSsl/CustomBio.cs

```diff
@@ -0,0 +1,253 @@
+        {
+            var bufferHandle = GCHandle.FromIntPtr((IntPtr)bio.ptr);
+            var bufferPipeline = (IPipelineWriter)bufferHandle.Target;
+            int numberOfBytesRemaing = numberOfBytes;
+            var buffer = bufferPipeline.Alloc();
+            while (numberOfBytesRemaing > 0)
+            {
+                int sizeToWrite = Math.Min(MaxBlockSize, numberOfBytesRemaing);
+                buffer.Ensure(sizeToWrite);
+                var writeSpan = new Span<byte>(buff, sizeToWrite);
+                writeSpan.CopyTo(buffer.Memory.Span);
+                buffer.Advance(sizeToWrite);
+                numberOfBytesRemaing -= sizeToWrite;
+            }
+            buffer.Commit();
+            bio.num_write = (uint)numberOfBytes;
+            BIO_set_flags(ref bio, BioFlags.BIO_FLAGS_WRITE | BioFlags.BIO_FLAGS_SHOULD_RETRY);
+            return numberOfBytes;
+        }
+
+        public static int NumberOfWrittenBytes(InteropBio.BioHandle handle)
+        {
+            var bio = (bio_st*)handle.Handle;
+            var bytes = (int)bio[0].num_write;
+            bio[0].num_write = 0;
+            return bytes;
+        }
+
+        private static int ReadBio(ref bio_st bio, void* buff, int numberOfBytes)
+        {
+            var buffer = Unsafe.Read<ReadableBuffer>(bio.ptr);
+            if (buffer.Length == 0)
+            {
+                bio.num = 0;
+                if (numberOfBytes > 0)
+                {
+                    BIO_set_flags(ref bio, BioFlags.BIO_FLAGS_READ | BioFlags.BIO_FLAGS_SHOULD_RETRY);
+                }
+                return -1;
+            }
+            if (numberOfBytes == 0)
+            {
+                return 0;
+            }
+            numberOfBytes = Math.Min(numberOfBytes, buffer.Length);
+            buffer.Slice(0, numberOfBytes).CopyTo(new Span<byte>(buff, numberOfBytes));
+            Unsafe.Write(bio.ptr, buffer.Slice(numberOfBytes));
+            bio.num = buffer.Length - numberOfBytes;
+            return numberOfBytes;
+        }
+
+        public static void SetReadBufferPointer(InteropBio.BioHandle bio, ref ReadableBuffer buffer)
+        {
+            var b = (bio_st*)bio.Handle;
+            b[0].num = buffer.Length;
+            b[0].ptr = Unsafe.AsPointer(ref buffer);
+        }
+
+        public static void SetWriteBufferPointer(InteropBio.BioHandle bio, GCHandle handleToWriteBuffer)
+        {
+            var b = (bio_st*)bio.Handle;
+            b[0].ptr = (void*)GCHandle.ToIntPtr(handleToWriteBuffer);
+        }
+
+        private static long ControlBio(ref bio_st bio, BioControl cmd, long num, void* ptr)
+        {
+            switch (cmd)
+            {
+                case BioControl.BIO_CTRL_FLUSH:
+                case BioControl.BIO_CTRL_POP:
+                case BioControl.BIO_CTRL_PUSH:
+                    return 1;
+            }
+            return 0;
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
```

Comment:
Maybe, maybe it can be visited in the the interop discussions. Might solve the problem i discuss below